lds/arm/task-rlink.x

```diff
@@ -84,4 +84,30 @@ SECTIONS
     *(.ARM.exidx.*);
     *(.ARM.extab.*);
   }
+
+  .debug_loc (INFO) : { *(.debug_loc) }
+  .debug_abbrev (INFO) : { *(.debug_abbrev) }
+  .debug_info (INFO) : { *(.debug_info) }
+  .debug_aranges (INFO) : { *(.debug_aranges) }
+  .debug_ranges (INFO) : { *(.debug_ranges) }
+  .debug_str (INFO) : { *(.debug_str) }
+  .debug_pubnames (INFO) : { *(.debug_pubnames) }
+  .debug_pubtypes (INFO) : { *(.debug_pubtypes) }
+  .debug_line (INFO) : { *(.debug_line) }
+  .debug_frame(INFO) : { *(.debug_frame) }
+
+  .llvmbc (INFO) : { *(.llvmbc) }
+  .llvmcmd (INFO) : { *(.llvmcmd) }
+
+  .note.GNU-stack (INFO) : { *(.note.GNU-stack) }
+
+  .ARM.attributes (INFO) : { *(.ARM.attributes) }
+
+  .symtab (INFO) : { *(.symtab) }
+  .shstrtab (INFO) : { *(.shstrtab) }
+  .strtab (INFO) : { *(.strtab) }
+
+  .eth_bulk (INFO) : { *(.eth_bulk) }
```

Comment:
When building `demo-stm32h7-nucleo/app-h753.toml` I get the following warning: `rust-lld: warning: ignoring memory region assignment for non-allocatable section '.eth_bulk'`. However, when I remove this line the linker complains about the section being orphaned. Is there something I can change about this to get rid of the warning?

Reply:
Isn't it wrong to mark `.eth_bulk` as `INFO`? This section needs to be allocated memory and marking it as `INFO` is incorrect for it.

Reply:
I have no idea ¯\_(ツ)_/¯. I tried looking up what that section is and couldn't find anything. I'm about to grep through `lld` source and see if that gets me somewhere.

Reply:
Oops. I missed linking this in my previous comment. [This](https://github.com/rivosinc/hubris/blob/rivos/main/task/net/src/buf.rs#L17) is where `.eth_bulk` is used.

Reply:
Actually, looks like the `net` task uses it for some ethernet data. Yeah, it should be allocated memory. Do you know what the line should be then?

Reply:
https://sourceware.org/binutils/docs/ld/Output-Section-Type.html

"These type names are supported for backward compatibility, and are rarely used. They all have the same effect: the section should be marked as not allocatable, so that no memory is allocated for the section when the program is run."

Is it .bss or initialized data? If it's initialized data it will need to be in the data section.

Reply:
Dropping the `(INFO)` should get rid of the warning:

```
diff --git a/lds/arm/task-rlink.x b/lds/arm/task-rlink.x
index 09a826b8..b97fea74 100644
--- a/lds/arm/task-rlink.x
+++ b/lds/arm/task-rlink.x
@@ -107,7 +107,7 @@ SECTIONS
   .shstrtab (INFO) : { *(.shstrtab) }
   .strtab (INFO) : { *(.strtab) }
 
-  .eth_bulk (INFO) : { *(.eth_bulk) }
+  .eth_bulk : { *(.eth_bulk) }
 
   .comment (INFO) : { *(.comment) }
 }

```

Reply:
> Is it .bss or initialized data? If it's initialized data it will need to be in the data section.

Not sure why the task defines a separate section for this. I am guessing if it is filled outside of the build somehow, then we might need that section, else I think we can go with .bss/.data depending upon whether it is initialized data or not. From a quick glance, it looks like the section contains ring buffers, so shouldn't really have any initialized data?

Reply:
Looking at the source, it appears as though all of these buffers should be zeroed, so I think we can put this in BSS. There is a [humility net](https://github.com/oxidecomputer/humility#humility-net) command, so maybe that's why it's separate?

Reply:
Probably? I had grepped `eth_bulk` in humility but didn't really find anything there.

Reply:
I just cloned upstream and grepped too, got this:
```
tests/cmd/extract/extract.ouray.22.stdout
168:sections = {eth_bulk = "sram1"}

tests/cmd/extract/extract.static-tasks.1.stdout
156:sections = {eth_bulk = "sram1"}

tests/cmd/extract/extract.ouray.21.stdout
168:sections = {eth_bulk = "sram1"}

tests/cmd/extract/extract.ouray.25.stdout
168:sections = {eth_bulk = "sram1"}

tests/cmd/extract/extract.ouray.24.stdout
168:sections = {eth_bulk = "sram1"}

tests/cmd/extract/extract.spoopy.0.stdout
66:sections = {eth_bulk = "sram1"}

tests/cmd/extract/extract.ouray.29.stdout
168:sections = {eth_bulk = "sram1"}

tests/cmd/extract/extract.chilly.0.stdout
66:sections = {eth_bulk = "sram1"}

tests/cmd/extract/extract.in_bootloader.0.stdout
41:sections = {eth_bulk = "sram1"}

tests/cmd/extract/extract.static-tasks.0.stdout
159:sections = {eth_bulk = "sram1"}

tests/cmd/extract/extract.kernel-panic.0.stdout
104:sections = {eth_bulk = "sram1"}

tests/cmd/extract/extract.ouray.26.stdout
168:sections = {eth_bulk = "sram1"}

tests/cmd/extract/extract.kiowa.26.stdout
116:sections = {eth_bulk = "sram1"}

tests/cmd/extract/extract.ouray.27.stdout
168:sections = {eth_bulk = "sram1"}

tests/cmd/extract/extract.ouray.30.stdout
168:sections = {eth_bulk = "sram1"}

tests/cmd/extract/extract.flash-ram-mismatch.0.stdout
41:sections = {eth_bulk = "sram1"}

tests/cmd/extract/extract.ouray.23.stdout
168:sections = {eth_bulk = "sram1"}
```

Not used in `net`, but it looks like they expect it to go into ram. I'm going to guess this is `.bss` as we're not using this task so it won't be a problem until we try and upstream.

Reply:
Aah, this goes in `app.toml`. Example on sidecar:
```
[tasks.net]
name = "task-net"
stacksize = 4640
priority = 5
features = ["mgmt", "h753", "sidecar", "vlan"]
max-sizes = {flash = 131072, ram = 16384, sram1 = 16384}
sections = {eth_bulk = "sram1"}
uses = ["eth", "eth_dma", "system_flash"]
start = true
interrupts = {"eth.irq" = 0b1}
task-slots = ["sys",
              { spi_driver = "spi3_driver" },
              { seq = "sequencer" }]
```

Looks like they are placing certain sections in different SRAM probably because it is faster? This mapping of sections to memory regions seems to be handled using [xtask](https://github.com/rivosinc/hubris/blob/rivos/main/build/xtask/src/dist.rs#L1007)

Reply:
This is what gets generated for `net` task for sidecar:

```
MEMORY
{
FLASH (rwx) : ORIGIN = 0x08020000, LENGTH = 0x00020000
STACK (rwx) : ORIGIN = 0x20004000, LENGTH = 0x00001220
RAM (rwx) : ORIGIN = 0x20005220, LENGTH = 0x00002de0
SRAM1 (rwx) : ORIGIN = 0x30000000, LENGTH = 0x00004000
}
SECTIONS {
  .eth_bulk (NOLOAD) : ALIGN(4) {
    *(.eth_bulk .eth_bulk.*);
  } > SRAM1
} INSERT AFTER .uninit
```

Reply:
Do you know which linker script(s) this is being added to?

Reply:
The linker script is generated by 'xtask` when linking the task into the final image. It gets overwritten after each task linking since it is placed in `target/memory.x`. You can use [this change](https://github.com/rivosinc/hubris/pull/33/commits/ae8a502aee4f4daeda1c6ecb166e5f62385341d0) to retain the generated linker scripts to inspect later. It should be named `target/sidecar/dist/net-memory.x` for sidecar after running `cargo xtask dist app/sidecar/app.toml`.

Reply:
I meant which of `task-link.x` and `task-tlink.x` this was being added to, although I think the `memory.x` nay be the same for both.

Reply:
As per [my reading](https://github.com/rivosinc/hubris/blob/rivos/main/build/xtask/src/dist.rs#L718), this is used with `task-link.x`.